src/rdrand.rs

```diff
@@ -9,33 +9,61 @@
 //! Implementation for SGX using RDRAND instruction
 use crate::Error;
 use core::mem;
-use core::arch::x86_64::_rdrand64_step;
+use core::arch::x86_64::{__cpuid, _rdrand64_step};
 use core::num::NonZeroU32;
-
-#[cfg(not(target_feature = "rdrand"))]
-compile_error!("enable rdrand target feature!");
+use lazy_static::lazy_static;
 
 // Recommendation from "Intel® Digital Random Number Generator (DRNG) Software
 // Implementation Guide" - Section 5.2.1 and "Intel® 64 and IA-32 Architectures
 // Software Developer’s Manual" - Volume 1 - Section 7.3.17.1.
 const RETRY_LIMIT: usize = 10;
 const WORD_SIZE: usize = mem::size_of::<u64>();
 
-fn rdrand() -> Result<[u8; WORD_SIZE], Error> {
+#[target_feature(enable = "rdrand")]
+unsafe fn rdrand() -> Result<[u8; WORD_SIZE], Error> {
     for _ in 0..RETRY_LIMIT {
-        unsafe {
-            // SAFETY: we've checked RDRAND support, and u64 can have any value.
-            let mut el = mem::uninitialized();
-            if _rdrand64_step(&mut el) == 1 {
-                return Ok(el.to_ne_bytes());
-            }
-        };
+        let mut el = mem::uninitialized();
```

Comment:
A bit late to the party here, but I was curious why this uses `mem::uninitialized()` instead of `mem::zeroed()`.

Zeroing seems relatively lightweight compared to RDRAND itself, would make RNG failures more obvious, and eliminates the potential risk of using attacker-controlled values which appear plausibly random at first glance as e.g. cryptographic keys.

Reply:
Yeah, `zeroed` could be a better choice. Although in practice it does not matter, as both `zeroed` and `uninitialized` get [optimized out](https://rust.godbolt.org/z/tBks1s).

Reply:
Fixed in the latest [commit](https://github.com/rust-random/getrandom/commit/1705b40655e99a3b3f974013ef31de9cd880864a).

Reply:
@tarcieri good catch! It also gets rid of a deprecation warning on the latest nightly.